app/models/candidate_interface/gcse_qualification_details_form.rb

```diff
@@ -17,16 +18,39 @@ def self.build_from_qualification(qualification)
     def save_base
       return false unless valid?
 
-      qualification = ApplicationQualification.find(qualification_id)
-
       qualification.update(grade: grade, award_year: award_year)
     end
 
   private
-
     def validate_award_year_is_date
       valid_award_year = award_year.match(/^[1-9]\d{3}$/)
       errors.add(:award_year, :invalid) unless valid_award_year
     end
+
+    def validate_grade_format
+      return if new_record? || qualification.qualification_type.nil?
+
+      qualification_rexp = invalid_grades[qualification.qualification_type.to_sym]
+
+      if qualification_rexp && grade.match(qualification_rexp)
+        errors.add(:grade, :invalid)
```

Comment:
Can we please add logging here that would:

- allow us to see which combination of qualification type and grade triggered the error
- allow us to easily grep for it in the logs

Reply:
Let's pull this out into a separate card, we'd want a generic solution: https://trello.com/c/zlfQ2il2/358-keep-track-of-validation-errors-in-the-application-form